core/unit_test/TestSharedHostPinnedSpace.cpp

```diff
@@ -0,0 +1,141 @@
+
+template <typename ExecutionSpace, typename ViewType>
+struct Increment {
+  ViewType view_;
+
+  explicit Increment(ExecutionSpace, ViewType view) : view_(view) {
+    Kokkos::fence();
+    Kokkos::parallel_for(
+        "increment",
+        Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>{
+            0, view_.size()},
+        *this);
+    Kokkos::fence();
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const size_t idx) const { ++view_(idx); }
+};
+
+template <typename ExecutionSpace, typename ViewType>
+struct CheckResult {
+  ViewType view_;
+  const int targetVal_;
+  unsigned error = 0;
+
+  CheckResult(ExecutionSpace, ViewType view, int targetVal)
+      : view_(view), targetVal_(targetVal) {
+    Kokkos::parallel_reduce(
+        "check",
+        Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>{
+            0, view_.size()},
+        *this, Kokkos::Sum<unsigned>(error));
+    Kokkos::fence();
+  }
+
+  KOKKOS_FUNCTION
+  void operator()(const size_t idx, unsigned& error_val) const {
+    if (view_(idx) != targetVal_) ++error_val;
+  }
+};
+
+TEST(defaultdevicetype, shared_host_pinned_space) {
+  ASSERT_TRUE(KOKKOS_HAS_SHARED_HOST_PINNED_SPACE);
+  ASSERT_TRUE(Kokkos::has_shared_host_pinned_space);
+
+  if constexpr (std::is_same_v<Kokkos::DefaultExecutionSpace,
+                               Kokkos::DefaultHostExecutionSpace>)
+    GTEST_SKIP() << "Skipping as host and device are the same space";
+
+  const unsigned int numDeviceHostCycles = 3;
+  size_t numInts                         = 1024;
+
+  // ALLOCATION
+  Kokkos::View<int*, Kokkos::SharedHostPinnedSpace> sharedData("sharedData",
+                                                               numInts);
+
+  // ASSESS RESULTS
+  bool readWriteAccessible = true;
+  unsigned incrementCount  = 0;
+
+  for (unsigned i = 0; i < numDeviceHostCycles; ++i) {
+    // INCREMENT DEVICE
+    Increment incrementDevice(Kokkos::DefaultExecutionSpace{}, sharedData);
+    ++incrementCount;
+    // CHECK RESULTS HOST
+    CheckResult checkHost(Kokkos::DefaultHostExecutionSpace{}, sharedData,
+                          incrementCount);
+    if (checkHost.error != 0) readWriteAccessible = false;
+
+    // INCREMENT HOST
+    Increment incrementHost(Kokkos::DefaultHostExecutionSpace{}, sharedData);
+    ++incrementCount;
+    // CHECK RESULTS Device
+    CheckResult checkDevice(Kokkos::DefaultExecutionSpace{}, sharedData,
+                            incrementCount);
+    if (checkDevice.error != 0) readWriteAccessible = false;
+  }
+
+  // CHECK IF PASSED
+  bool passed = (readWriteAccessible);
+
+  // PRINT IF NOT PASSED
+  if (!passed) {
+    std::cout << "Allocating " << numInts
+              << " ints in SharedHostPinnedSpace.\n";
+
+    std::cout << "Behavior found: \n";
+    std::cout
+        << "SharedHostPinnedSpace is read/write accessible by host and device: "
+        << readWriteAccessible << ", we expect true \n\n";
+  }
+  ASSERT_TRUE(passed);
```

Comment:
```suggestion
  ASSERT_TRUE(readWriteAccessible) << "PRINT ON FAILURE";
```

Reply:
done in [d97863](https://github.com/kokkos/kokkos/pull/5405/commits/d9786320d3f5711f8b339c39bcf0e854fa5b64c1)